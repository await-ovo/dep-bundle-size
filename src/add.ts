import { spawn } from 'node:child_process';
import { spinner } from '@clack/prompts';
import { bold } from 'kolorist';
import {
  detectPackageManager,
  findRootDirectory,
  formatSize,
  getConfig,
  logger,
  requestStats,
} from './utils';
import type { PackageStatsResponse, BundlephobiaConfig } from './utils';

const checkConstraints = ({
  constraints,
  config,
  stats,
  warning,
}: {
  constraints: { key: keyof BundlephobiaConfig; name: string }[];
  stats: PackageStatsResponse[];
  config: BundlephobiaConfig;
  warning?: boolean;
}) => {
  for (const constraint of constraints) {
    const limit = config[constraint.key];
    if (!limit) {
      continue;
    }
    const exceeded = stats.filter(({ size }) => size > limit);

    const hint = `Can not install ${exceeded
      .map(({ name, version, size }) =>
        bold(`${name}@${version}(${formatSize(size)})`),
      )
      .join(',')}, Since their ${
      constraint.name
    } is larger than the configured(${bold(formatSize(limit))}).`;
    if (warning) {
      logger.warn(hint);
    } else {
      throw new Error(hint);
    }
  }

  const { overallSize, overallGzipSize } = stats.reduce<{
    overallSize: number;
    overallGzipSize: number;
  }>(
    (sizes, { size, gzip }) => {
      sizes.overallSize += size;
      sizes.overallGzipSize += gzip;
      return sizes;
    },
    {
      overallSize: 0,
      overallGzipSize: 0,
    },
  );

  if (config.maxOverallSize && overallSize > config.maxOverallSize) {
    const hint = `Can not install ${stats
      .map(({ name, version, size }) =>
        bold(`${name}@${version}(${formatSize(size)})`),
      )
      .join(',')}, Since their total size is larger than the configured(${bold(
      formatSize(config.maxOverallSize),
    )}).`;

    if (warning) {
      logger.warn(hint);
    } else {
      throw new Error(hint);
    }
  }

  if (config.maxGzipSize && overallGzipSize > config.maxGzipSize) {
    const hint = `Can not install ${stats
      .map(
        ({ name, version, gzip }) =>
          `${name}@${version}(${bold(formatSize(gzip))})`,
      )
      .join(',')}, Since their total size is larger than the configured(${bold(
      formatSize(config.maxGzipSize),
    )}).`;
    if (warning) {
      logger.warn(hint);
    } else {
      throw new Error(hint);
    }
  }
};

const doAdd = async ({
  packages,
  cwd,
  rootProjectDir,
  dryRun,
}: {
  packages: string[];
  cwd: string;
  rootProjectDir: string;
  dryRun?: boolean;
}) => {
  const packageManager = detectPackageManager(rootProjectDir);

  const args = [
    packageManager,
    packageManager === 'npm' ? 'install' : 'add',
    ...packages,
  ];
  logger.log(`Add deps(${bold(args.join(' '))})`);
  if (!dryRun) {
    spawn(args[0], args.slice(1), {
      cwd,
      stdio: 'inherit',
    });
  }
};

export const add = async (
  packages: string[],
  options: {
    warning?: boolean;
    dir?: string;
    dryRun?: boolean;
  },
) => {
  const checkSpinner = spinner();
  checkSpinner.start('Checking constraints');
  const rootProjectDir = findRootDirectory(options.dir ?? process.cwd());
  const config = await getConfig(rootProjectDir);

  const stats = await requestStats(packages);

  checkConstraints({
    constraints: [
      { name: 'max-size', key: 'maxSize' },
      { name: 'max-gzip-size', key: 'maxGzipSize' },
    ],
    config,
    warning: options.warning,
    stats,
  });

  checkSpinner.stop('Checking constraints completed');

  await doAdd({
    packages,
    cwd: options.dir ?? process.cwd(),
    rootProjectDir,
    dryRun: options.dryRun,
  });
};
