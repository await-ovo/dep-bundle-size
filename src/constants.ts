export const BUNDLE_PHOBIA_API = 'https://bundlephobia.com/api/size';

export enum ErrorType {
  ParseJSONError,
}

export const ERROR_MESSAGES = {
  [ErrorType.ParseJSONError]: ({ file }: { file: string }) =>
    `JSON parse failed. file: ${file}`,
};
