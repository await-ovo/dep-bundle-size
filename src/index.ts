import path from 'node:path';
import os from 'os';
import enquirer from 'enquirer';
import { table } from 'table';
import pLimit from 'p-limit';
import ora from 'ora';
import {
  findWorkspaceProjects,
  formatSize,
  formatTime,
  getDependencies,
  getDepInstalledVersion,
  getTimeFromSize,
  logger,
  requestStats,
  RuntimeError,
} from './utils';
import { ErrorType } from './constants';
import type { Project } from './utils';

const { prompt } = enquirer;

const doScan = async ({
  project,
  projects,
  rootProjectDir,
  packageArgs,
  interactive,
}: {
  project: Project;
  projects: Project[];
  rootProjectDir: string;
  packageArgs?: string[];
  interactive?: boolean;
}) => {
  const { path: projectDir } = project;
  const relativeProjectPath = path.relative(rootProjectDir, projectDir);
  const relativeName = relativeProjectPath === '' ? './' : relativeProjectPath;

  const spinner = ora({
    text: `Project ${relativeName} ...`,
    color: 'yellow',
  }).start();

  const deps = getDependencies(path.join(projectDir, 'package.json'), projects);

  let packages = packageArgs?.length ? packageArgs : Object.keys(deps);

  if (interactive) {
    ({ packages } = await prompt({
      type: 'multiselect',
      name: 'packages',
      message: 'Select packages to scan?',
      choices: packages,
    }));
  }

  if (!packages.length) {
    logger.warn(`No Packages to scan in project: ${projectDir}`);
    spinner.fail();
    return;
  }

  const stats = await requestStats(
    packages.map(
      packageName =>
        `${packageName}@${
          getDepInstalledVersion(packageName, projectDir) ??
          deps[packageName] ??
          'latest'
        }`,
    ),
  );

  spinner.text = `Project ${relativeName} completed:`;
  spinner.succeed();

  /* eslint-disable no-console */
  console.log(
    table(
      [
        ['Name', 'MIN', 'MIN + GZIP', 'SLOW 3G', 'EMERGING 4G'],
        ...stats.map(record => {
          const time = getTimeFromSize(record.gzip);
          return [
            `${record.name}@${record.version}`,
            formatSize(record.size),
            formatSize(record.gzip),
            formatTime(time.threeG),
            formatTime(time.fourG),
          ];
        }),
      ],
      {
        header: {
          alignment: 'center',
          content: 'Scan Results',
        },
      },
    ),
  );
  /* eslint-enable no-console */
};

export const scan = async (
  packageArgs: string[],
  options: { interactive?: boolean; recursive?: boolean },
) => {
  if (options.recursive && options.interactive) {
    throw new RuntimeError(ErrorType.RecursiveWithInteractiveError);
  }

  const rootProjectDir = process.cwd();

  const projects: Project[] = [
    {
      name: '.',
      path: rootProjectDir,
    },
  ];

  if (options.recursive) {
    projects.push(...(await findWorkspaceProjects(rootProjectDir)));
  }

  const limit = pLimit(os.cpus().length);

  await Promise.all(
    projects.map(project =>
      limit(() =>
        doScan({
          project,
          rootProjectDir,
          projects,
          packageArgs,
          interactive: options.interactive,
        }),
      ),
    ),
  );
};
