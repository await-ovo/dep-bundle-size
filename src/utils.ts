import fs from 'node:fs';
import path from 'node:path';
import { got, HTTPError } from 'got';
import pLimit from 'p-limit';
import yaml from 'js-yaml';
import glob from 'fast-glob';
import { BUNDLE_PHOBIA_API, ErrorType, NPM_REGISTRY } from './constants';

const limit = pLimit(6);

export type Project = {
  name?: string;
  path: string;
};

export type PackageStatsResponse = {
  assets: {
    gzip: number;
    name: string;
    size: number;
    type: string;
  }[];
  dependencyCount: number;
  dependencySizes: {
    approximateSize: number;
    name: string;
  }[];
  description: string;
  gzip: number;
  hasJSModule: string;
  hasJSNext: string;
  hasSideEffects: string[];
  name: string;
  isModuleType: boolean;
  repository: string;
  scoped: boolean;
  size: number;
  version: string;
};

export const logger = {
  log: (...args: any[]) => {
    process.stdout.write('\nINFO: ');
    // eslint-disable-next-line no-console
    console.log('[INFO]: ', ...args);
  },

  error: (...args: any[]) => {
    process.stderr.write('\nERROR: ');
    // eslint-disable-next-line no-console
    console.error(...args);
  },
  warn: (...args: any[]) => {
    process.stdout.write('\nWARN: ');
    // eslint-disable-next-line no-console
    console.warn(...args);
  },
};

export class RuntimeError extends Error {
  errorType: ErrorType;
  options?: any;
  constructor(errorType: ErrorType, options?: any) {
    super();
    this.errorType = errorType;
    this.options = options;
  }
}

export type PackageManager = 'pnpm' | 'yarn' | 'npm';

export const detectPackageManager = (dir: string) => {
  if (fs.existsSync(path.join(dir, 'pnpm-lock.yaml'))) {
    return 'pnpm';
  } else if (fs.existsSync(path.join(dir, 'package-lock.json'))) {
    return 'npm';
  } else if (fs.existsSync(path.join(dir, 'yarn.lock'))) {
    return 'yarn';
  }

  return 'npm';
};

export const getDepInstalledVersion = (
  specifier: string,
  dir?: string,
): string | undefined => {
  try {
    const packagePath = path.join(
      dir ?? process.cwd(),
      `node_modules/${specifier}/package.json`,
    );

    return JSON.parse(fs.readFileSync(packagePath, 'utf8')).version;
  } catch (err) {
    logger.warn(
      `Can not get installed version of package ${specifier}, use version listed in package.json or latest instead`,
    );
  }
};

export const getDependencies = (file: string, projects: Project[]) => {
  try {
    const json = JSON.parse(fs.readFileSync(file, 'utf8'));
    const dependencies = json.dependencies ?? {};

    // remove local dependencies
    return Object.keys(dependencies)
      .filter(depName => !projects.find(project => project.name === depName))
      .reduce<Record<string, string>>((deps, depName) => {
        deps[depName] = dependencies[depName];
        return deps;
      }, {});
  } catch (err) {
    throw new RuntimeError(ErrorType.ParseJSONError, { file });
  }
};

const memoize = <T>(fn: (...args: string[]) => Promise<T>) => {
  const cached = new Map<string, T>();

  return async (...args: string[]) => {
    const key = args.join('_');
    if (!cached.has(key)) {
      const res = await fn(...args);
      cached.set(key, res);
      return res;
    } else {
      return cached.get(key)!;
    }
  };
};

const requestPackageStats = memoize<PackageStatsResponse>(
  async (specifier: string) => {
    try {
      const packageStats: PackageStatsResponse = await got(
        `${BUNDLE_PHOBIA_API}?package=${specifier}&record=true`,
        {
          headers: {
            'User-Agent': 'bundle-phobia-cli',
            'X-Bundlephobia-User': 'bundle-phobia-cli',
          },
        },
      ).json();
      return packageStats;
    } catch (err) {
      if (err instanceof HTTPError) {
        const { code, message } = JSON.parse(err.response.body as string).error;
        throw new RuntimeError(ErrorType.RequestStatsError, {
          specifier,
          code,
          message,
        });
      } else {
        throw err;
      }
    }
  },
);

export const requestStats = async (specifiers: string[]) =>
  Promise.all(
    specifiers.map(specifier => limit(() => requestPackageStats(specifier))),
  );

export const findWorkspaceProjects = async (
  dir: string,
): Promise<Project[]> => {
  const packageManager = detectPackageManager(dir);

  let workspacePatterns: string[] = [];

  // pnpm workspaces
  if (
    packageManager === 'pnpm' &&
    fs.existsSync(path.join(dir, 'pnpm-workspace.yaml'))
  ) {
    const yamlContent = yaml.load(
      fs.readFileSync(path.join(dir, 'pnpm-workspace.yaml'), 'utf8'),
    ) as { packages?: string[] };
    workspacePatterns = yamlContent.packages ?? [];
  } else {
    const json = JSON.parse(
      fs.readFileSync(path.join(dir, 'package.json'), 'utf8'),
    );

    workspacePatterns = json.workspaces ?? [];
  }

  return workspacePatterns?.length
    ? (
        await glob(workspacePatterns, {
          onlyDirectories: true,
          unique: true,
          absolute: true,
        })
      )
        .filter(matched => fs.existsSync(path.join(matched, 'package.json')))
        .map(directory => ({
          name: (
            JSON.parse(
              fs.readFileSync(path.join(directory, 'package.json'), 'utf8'),
            ) as { name?: string }
          ).name,
          path: directory,
        }))
    : [];
};

// To be consistent with bundlephobia Web UI, the formatTime and getTimeFromSize utility functions modified from:
// https://github.com/pastelsky/bundlephobia/blob/bundlephobia/utils/index.js#L27
export const formatSize = (value: number) => {
  let unit, size;
  if (Math.log10(value) < 3) {
    unit = 'B';
    size = value;
  } else if (Math.log10(value) < 6) {
    unit = 'kB';
    size = value / 1024;
  } else {
    unit = 'MB';
    size = value / 1024 / 1024;
  }

  return `${size.toFixed(2)}${unit}`;
};

export const formatTime = (value: number) => {
  let unit, size;
  if (value < 0.0005) {
    unit = 'Î¼s';
    size = Math.round(value * 1000000);
  } else if (value < 0.5) {
    unit = 'ms';
    size = Math.round(value * 1000);
  } else {
    unit = 's';
    size = value;
  }

  return `${size}${unit}`;
};

export const getTimeFromSize = (sizeInBytes: number) => {
  return {
    threeG: sizeInBytes / 1024 / (400 / 8),
    fourG: sizeInBytes / 1024 / (7000 / 8),
  };
};
