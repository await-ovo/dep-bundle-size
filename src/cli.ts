#!/usr/bin/env node
import { cac } from 'cac';
import { scan } from '.';
import { add } from './add';
import { logger, RuntimeError } from './utils';
import { ERROR_MESSAGES } from './constants';
import { version } from '../package.json';

const errorHandler = (err: any) => {
  if (err instanceof RuntimeError) {
    logger.error(ERROR_MESSAGES[err.errorType](err.options), '\n', err.stack);
  } else {
    logger.error('%o', err);
  }
  process.exit(1);
};

process.on('unhandledRejection', errorHandler);
process.on('uncaughtException', errorHandler);

const cli = cac('dep-bundle-size');

cli.help();
cli.version(version);

cli
  .command('[...packages]')
  .option('-i, --interactive', 'Select packages to scan')
  .option('-r, --recursive', 'Select packages to scan')
  .option('-d, --dir [dir]', 'working directory')
  .action(scan);

cli
  .command('add <...packages>')
  .option('-w, --warning', 'use warning instead error')
  .option('-d, --dir [dir]', 'working directory')
  .action(add);

cli.parse();
