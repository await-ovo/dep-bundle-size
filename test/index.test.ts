import path from 'node:path';
import { expect, test, vi } from 'vitest';
import { scan } from '../src';

const fixtures = path.join(process.cwd(), 'test/fixtures');

const createConsoleMock = () => {
  let output = '';
  return {
    startMock: () => {
      vi.stubGlobal('console', {
        log: vi.fn((...args: any[]) => (output += args.join(' '))),
        error: vi.fn((...args: any[]) => (output += args.join(' '))),
        warn: vi.fn((...args: any[]) => (output += args.join(' '))),
      });
    },
    stopMock: () => {
      vi.unstubAllGlobals();
    },
    getOutput: () => output,
  };
};

test('auto scan deps in single project', async () => {
  const { getOutput, startMock, stopMock } = createConsoleMock();
  startMock();
  await scan([], {
    dir: path.join(fixtures, 'base-project'),
  });
  stopMock();
  expect(getOutput()).toMatchSnapshot();
});

test('specify packages to scan in single project', async () => {
  const { getOutput, startMock, stopMock } = createConsoleMock();
  startMock();
  await scan(['react', 'dayjs'], {
    dir: path.join(fixtures, 'base-project'),
  });
  stopMock();
  expect(getOutput()).toMatchSnapshot();
});

test('recursive scan deps in pnpm workspaces', async () => {
  const { getOutput, startMock, stopMock } = createConsoleMock();
  startMock();
  await scan([], {
    dir: path.join(fixtures, 'pnpm-workspace'),
    recursive: true,
    workspaceConcurrency: 1,
  });
  stopMock();
  expect(getOutput()).toMatchSnapshot();
});

test('specify packages to scan in all packages', async () => {
  const { getOutput, startMock, stopMock } = createConsoleMock();
  startMock();
  await scan(['react', 'dayjs'], {
    dir: path.join(fixtures, 'pnpm-workspace'),
    recursive: true,
    workspaceConcurrency: 1,
  });
  stopMock();
  expect(getOutput()).toMatchSnapshot();
});

test('recursive scan deps in yarn workspaces', async () => {
  const { getOutput, startMock, stopMock } = createConsoleMock();
  startMock();
  await scan([], {
    dir: path.join(fixtures, 'yarn-workspace'),
    recursive: true,
    workspaceConcurrency: 1,
  });
  stopMock();
  expect(getOutput()).toMatchSnapshot();
});
