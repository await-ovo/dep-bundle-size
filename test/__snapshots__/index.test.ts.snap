// Vitest Snapshot v1

exports[`auto scan deps in single project 1`] = `
"╔════════════════════════════════════════════════════════════════════════╗
║                              Scan Results                              ║
╟──────────────────┬──────────┬────────────┬───────────────┬─────────────╢
║ Name             │ MIN      │ MIN + GZIP │ SLOW 3G       │ EMERGING 4G ║
╟──────────────────┼──────────┼────────────┼───────────────┼─────────────╢
║ react@18.2.0     │ 6.43kB   │ 2.51kB     │ 50ms          │ 3ms         ║
╟──────────────────┼──────────┼────────────┼───────────────┼─────────────╢
║ react-dom@18.2.0 │ 130.48kB │ 41.99kB    │ 0.8398046875s │ 48ms        ║
╟──────────────────┼──────────┼────────────┼───────────────┼─────────────╢
║ classnames@2.3.2 │ 717.00B  │ 431.00B    │ 8ms           │ 481μs       ║
╚══════════════════╧══════════╧════════════╧═══════════════╧═════════════╝
 
"
`;

exports[`recursive scan deps in pnpm workspaces 1`] = `
"╔═════════════════════════════════════════════════════════════════════╗
║                            Scan Results                             ║
╟──────────────────────┬─────────┬────────────┬─────────┬─────────────╢
║ Name                 │ MIN     │ MIN + GZIP │ SLOW 3G │ EMERGING 4G ║
╟──────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ react-intl@6.2.8     │ 62.89kB │ 17.79kB    │ 356ms   │ 20ms        ║
╟──────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ react-i18next@12.1.5 │ 23.32kB │ 7.08kB     │ 142ms   │ 8ms         ║
╟──────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ eo-locale@7.4.2      │ 5.17kB  │ 1.79kB     │ 36ms    │ 2ms         ║
╚══════════════════════╧═════════╧════════════╧═════════╧═════════════╝
 
╔══════════════════════════════════════════════════════════════════════════════════════╗
║                                     Scan Results                                     ║
╟───────────────────────────────┬──────────┬────────────┬────────────────┬─────────────╢
║ Name                          │ MIN      │ MIN + GZIP │ SLOW 3G        │ EMERGING 4G ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ lodash@4.17.21                │ 69.89kB  │ 24.47kB    │ 489ms          │ 28ms        ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ axios@0.27.2                  │ 19.20kB  │ 6.74kB     │ 135ms          │ 8ms         ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ axios-ntlm@1.3.0              │ 28.22kB  │ 9.68kB     │ 194ms          │ 11ms        ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ badge-maker@3.3.1             │ 903.43kB │ 286.25kB   │ 5.72494140625s │ 327ms       ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ bcryptjs@2.4.3                │ 21.24kB  │ 9.59kB     │ 192ms          │ 11ms        ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ bree@7.1.5                    │ 157.03kB │ 45.99kB    │ 0.91970703125s │ 53ms        ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ cacheable-lookup@6.0.4        │ 4.90kB   │ 1.92kB     │ 38ms           │ 2ms         ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ chardet@1.4.0                 │ 35.13kB  │ 9.64kB     │ 193ms          │ 11ms        ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ check-password-strength@2.0.7 │ 1.03kB   │ 592.00B    │ 12ms           │ 1ms         ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ dayjs@1.11.7                  │ 6.46kB   │ 2.88kB     │ 58ms           │ 3ms         ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ form-data@4.0.0               │ 203.00B  │ 169.00B    │ 3ms            │ 189μs       ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ gamedig@4.0.6                 │ 290.95kB │ 161.08kB   │ 3.221640625s   │ 184ms       ║
╚═══════════════════════════════╧══════════╧════════════╧════════════════╧═════════════╝
 
╔══════════════════════════════════════════════════════════════════════╗
║                             Scan Results                             ║
╟───────────────────────┬─────────┬────────────┬─────────┬─────────────╢
║ Name                  │ MIN     │ MIN + GZIP │ SLOW 3G │ EMERGING 4G ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ lodash@4.17.21        │ 69.89kB │ 24.47kB    │ 489ms   │ 28ms        ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ axios@0.27.2          │ 19.20kB │ 6.74kB     │ 135ms   │ 8ms         ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ axios-ntlm@1.3.0      │ 28.22kB │ 9.68kB     │ 194ms   │ 11ms        ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ react-dropzone@7.0.1  │ 16.99kB │ 5.72kB     │ 114ms   │ 7ms         ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ react-flip-move@3.0.5 │ 14.89kB │ 4.84kB     │ 97ms    │ 6ms         ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ react-sidebar@3.0.2   │ 7.46kB  │ 2.38kB     │ 48ms    │ 3ms         ║
╚═══════════════════════╧═════════╧════════════╧═════════╧═════════════╝
 
"
`;

exports[`recursive scan deps in yarn workspaces 1`] = `
"Warn:  No Packages to scan in project: /Users/zhenwei.song/Documents/projects/dep-bundle-size/test/fixtures/yarn-workspace
╔══════════════════════════════════════════════════════════════════════════════════════╗
║                                     Scan Results                                     ║
╟───────────────────────────────┬──────────┬────────────┬────────────────┬─────────────╢
║ Name                          │ MIN      │ MIN + GZIP │ SLOW 3G        │ EMERGING 4G ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ lodash@4.17.21                │ 69.89kB  │ 24.47kB    │ 489ms          │ 28ms        ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ axios@0.27.2                  │ 19.20kB  │ 6.74kB     │ 135ms          │ 8ms         ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ axios-ntlm@1.3.0              │ 28.22kB  │ 9.68kB     │ 194ms          │ 11ms        ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ badge-maker@3.3.1             │ 903.43kB │ 286.25kB   │ 5.72494140625s │ 327ms       ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ bcryptjs@2.4.3                │ 21.24kB  │ 9.59kB     │ 192ms          │ 11ms        ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ bree@7.1.5                    │ 157.03kB │ 45.99kB    │ 0.91970703125s │ 53ms        ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ cacheable-lookup@6.0.4        │ 4.90kB   │ 1.92kB     │ 38ms           │ 2ms         ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ chardet@1.4.0                 │ 35.13kB  │ 9.64kB     │ 193ms          │ 11ms        ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ check-password-strength@2.0.7 │ 1.03kB   │ 592.00B    │ 12ms           │ 1ms         ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ dayjs@1.11.7                  │ 6.46kB   │ 2.88kB     │ 58ms           │ 3ms         ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ form-data@4.0.0               │ 203.00B  │ 169.00B    │ 3ms            │ 189μs       ║
╟───────────────────────────────┼──────────┼────────────┼────────────────┼─────────────╢
║ gamedig@4.0.6                 │ 290.95kB │ 161.08kB   │ 3.221640625s   │ 184ms       ║
╚═══════════════════════════════╧══════════╧════════════╧════════════════╧═════════════╝
 
╔══════════════════════════════════════════════════════════════════════╗
║                             Scan Results                             ║
╟───────────────────────┬─────────┬────────────┬─────────┬─────────────╢
║ Name                  │ MIN     │ MIN + GZIP │ SLOW 3G │ EMERGING 4G ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ lodash@3.10.1         │ 49.84kB │ 18.24kB    │ 365ms   │ 21ms        ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ axios@0.27.2          │ 19.20kB │ 6.74kB     │ 135ms   │ 8ms         ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ axios-ntlm@1.3.0      │ 28.22kB │ 9.68kB     │ 194ms   │ 11ms        ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ react-dropzone@7.0.1  │ 16.99kB │ 5.72kB     │ 114ms   │ 7ms         ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ react-flip-move@3.0.5 │ 14.89kB │ 4.84kB     │ 97ms    │ 6ms         ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ react-sidebar@3.0.2   │ 7.46kB  │ 2.38kB     │ 48ms    │ 3ms         ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ react-intl@6.2.8      │ 62.89kB │ 17.79kB    │ 356ms   │ 20ms        ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ react-i18next@12.1.5  │ 23.32kB │ 7.08kB     │ 142ms   │ 8ms         ║
╟───────────────────────┼─────────┼────────────┼─────────┼─────────────╢
║ eo-locale@7.4.2       │ 5.17kB  │ 1.79kB     │ 36ms    │ 2ms         ║
╚═══════════════════════╧═════════╧════════════╧═════════╧═════════════╝
 
Warn:  No Packages to scan in project: /Users/zhenwei.song/Documents/projects/dep-bundle-size/test/fixtures/yarn-workspace/packages/zoo
Warn:  No Packages to scan in project: /Users/zhenwei.song/Documents/projects/dep-bundle-size/test/fixtures/yarn-workspace/apps/native/lib
"
`;

exports[`specify packages to scan in all packages 1`] = `
"Warn:  Can not get installed version of package react, use version listed in package.json or latest instead
Warn:  Can not get installed version of package dayjs, use version listed in package.json or latest instead
╔════════════════════════════════════════════════════════════╗
║                        Scan Results                        ║
╟──────────────┬────────┬────────────┬─────────┬─────────────╢
║ Name         │ MIN    │ MIN + GZIP │ SLOW 3G │ EMERGING 4G ║
╟──────────────┼────────┼────────────┼─────────┼─────────────╢
║ react@18.2.0 │ 6.43kB │ 2.51kB     │ 50ms    │ 3ms         ║
╟──────────────┼────────┼────────────┼─────────┼─────────────╢
║ dayjs@1.11.7 │ 6.46kB │ 2.88kB     │ 58ms    │ 3ms         ║
╚══════════════╧════════╧════════════╧═════════╧═════════════╝
 
Warn:  Can not get installed version of package react, use version listed in package.json or latest instead
╔════════════════════════════════════════════════════════════╗
║                        Scan Results                        ║
╟──────────────┬────────┬────────────┬─────────┬─────────────╢
║ Name         │ MIN    │ MIN + GZIP │ SLOW 3G │ EMERGING 4G ║
╟──────────────┼────────┼────────────┼─────────┼─────────────╢
║ react@18.2.0 │ 6.43kB │ 2.51kB     │ 50ms    │ 3ms         ║
╟──────────────┼────────┼────────────┼─────────┼─────────────╢
║ dayjs@1.11.7 │ 6.46kB │ 2.88kB     │ 58ms    │ 3ms         ║
╚══════════════╧════════╧════════════╧═════════╧═════════════╝
 
Warn:  Can not get installed version of package react, use version listed in package.json or latest instead
Warn:  Can not get installed version of package dayjs, use version listed in package.json or latest instead
╔════════════════════════════════════════════════════════════╗
║                        Scan Results                        ║
╟──────────────┬────────┬────────────┬─────────┬─────────────╢
║ Name         │ MIN    │ MIN + GZIP │ SLOW 3G │ EMERGING 4G ║
╟──────────────┼────────┼────────────┼─────────┼─────────────╢
║ react@18.2.0 │ 6.43kB │ 2.51kB     │ 50ms    │ 3ms         ║
╟──────────────┼────────┼────────────┼─────────┼─────────────╢
║ dayjs@1.11.7 │ 6.46kB │ 2.88kB     │ 58ms    │ 3ms         ║
╚══════════════╧════════╧════════════╧═════════╧═════════════╝
 
"
`;

exports[`specify packages to scan in single project 1`] = `
"Warn:  Can not get installed version of package dayjs, use version listed in package.json or latest instead
╔════════════════════════════════════════════════════════════╗
║                        Scan Results                        ║
╟──────────────┬────────┬────────────┬─────────┬─────────────╢
║ Name         │ MIN    │ MIN + GZIP │ SLOW 3G │ EMERGING 4G ║
╟──────────────┼────────┼────────────┼─────────┼─────────────╢
║ react@18.2.0 │ 6.43kB │ 2.51kB     │ 50ms    │ 3ms         ║
╟──────────────┼────────┼────────────┼─────────┼─────────────╢
║ dayjs@1.11.7 │ 6.46kB │ 2.88kB     │ 58ms    │ 3ms         ║
╚══════════════╧════════╧════════════╧═════════╧═════════════╝
 
"
`;
